File: cs/example/SOEM.cs
/*
 * File: SOEM.cs
 * Project: example
 * Created Date: 30/04/2021
 * Author: Shun Suzuki
 * -----
 * Last Modified: 09/10/2022
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022 Shun Suzuki. All rights reserved.
 * 
 */

using AUTD3Sharp;
using System;
using AUTD3Sharp.Link;
using AUTD3Sharp.Utils;
using example.Test;

namespace example
{
    internal static class SOEMTest
    {
        public static void Test()
        {
            Console.WriteLine("Test with SOEM");

            var autd = new Controller();
            autd.AddDevice(Vector3d.Zero, Vector3d.Zero);
            //autd.AddDevice(Vector3d.Zero, Vector3d.Zero, 1);

            // Controller.ToNormal();
            // for (int i = 0; i < Controller.NumTransInDevice; i++)
            //     autd.SetTransFrequency(0, i, 70e3);

            var link = new SOEM()
                .HighPrecision(true)
                .OnLost(x =>
                    {
                        Console.WriteLine($"Unrecoverable error occurred: {x}");
                        Environment.Exit(-1);
                    })
                .Build();
            if (!autd.Open(link))
            {
                Console.WriteLine(Controller.LastError);
                return;
            }

            autd.CheckTrials = 50;

            TestRunner.Run(autd);
        }
    }
}

File: julia/example/soem.jl
# File: soem.jl
# Project: example
# Created Date: 30/12/2020
# Author: Shun Suzuki
# -----
# Last Modified: 18/04/2023
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2020 Hapis Lab. All rights reserved.
# 

using Printf

using AUTD3
using StaticArrays
using StaticArrays: size

include("tests/runner.jl")

function on_lost(msg::Cstring)
    println(msg)
    exit(-1)
end

function main()
    try
        geometry = GeometryBuilder().add_device(SVector(0.0, 0.0, 0.0), SVector(0.0, 0.0, 0.0)).build()

        link = SOEM().on_lost(on_lost).build()

        cnt = Controller(geometry, link)

        run(cnt)

    catch e
        println(e)
    end
end

main()
